<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="3"
      main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <TimingDecorator name="Total">
      <Sequence>
        <Sequence>
          <SetBlackboard value="-0.35,0.0,0.0"
                         output_key="pre_grab_box_offset"/>
          <SetBlackboard value="-0.35,0.0,0.0"
                         output_key="pre_grab_box_offset_after_sgl_step"/>
          <SetBlackboard value="0.35,0.4,0.3"
                         output_key="box_size"/>
          <SetBlackboard value="-0.4,0,0"
                         output_key="pre_grab_box_offset2"/>
        </Sequence>
        <CheckStatusOK/>
        <ArmMoveToReadyPose/>
        <SequenceStar>
          <SequenceStar>
            <TimingDecorator name="MoveToGraspCMD">
              <Sequence>
                <RetryUntilSuccessful num_attempts="3">
                  <BoxTagOK tag_id="1"
                            box_pos="{box_pose}"
                            box_quat="{box_quat}"/>
                </RetryUntilSuccessful>
                <ComputeTargetPose box_pos="{box_pose}"
                                   box_quat="{box_quat}"
                                   target_box_offset="{pre_grab_box_offset}"
                                   target_pose="{target_pose_pre_grasp}"/>
                <MakePlan target_pose="{target_pose_pre_grasp}"
                          planed_trajectory="{planed_trajectory}"/>
                <MoveToDestination planed_trajectory="{planed_trajectory}"
                                   move_type="0"/>
              </Sequence>
            </TimingDecorator>
            <TimingDecorator name="Grasp">
              <Sequence>
                <Sequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <BoxTagOK tag_id="1"
                              box_pos="{box_pose}"
                              box_quat="{box_quat}"/>
                  </RetryUntilSuccessful>
                  <ComputeTargetPose box_pos="{box_pose}"
                                     box_quat="{box_quat}"
                                     target_box_offset="{pre_grab_box_offset_after_sgl_step}"
                                     target_pose="{target_pose_pre_grasp}"/>
                </Sequence>
                <RetryUntilSuccessful num_attempts="3">
                  <GraspBox box_pos="{box_pose}"
                            box_quat="{box_quat}"
                            box_size="{box_size}"
                            box_offset="0.0,0.0,0.06"
                            grasp_type="0"/>
                </RetryUntilSuccessful>
              </Sequence>
            </TimingDecorator>
            <TimingDecorator name="GraspBcakCMD">
              <Sequence>
                <ComputeTargetPose box_pos="{box_pose}"
                                   box_quat="{box_quat}"
                                   target_box_offset="-0.7,0,0"
                                   target_pose="{target_pose_pre_put}"/>
                <MakePlan target_pose="{target_pose_pre_put}"
                          planed_trajectory="{planed_trajectory_pre_put}"/>
              </Sequence>
            </TimingDecorator>
          </SequenceStar>
          <SequenceStar>
            <TimingDecorator name="MoveToPutSingle(Turn)">
              <Sequence>
                <RetryUntilSuccessful num_attempts="3">
                  <BoxTagOK tag_id="2"
                            box_pos="{box_pose2}"
                            box_quat="{box_quat2}"/>
                </RetryUntilSuccessful>
                <ComputeTargetPose box_pos="{box_pose2}"
                                   box_quat="{box_quat2}"
                                   target_box_offset="{pre_grab_box_offset2}"
                                   target_pose="{target_pose_pre_put}"/>
                <ComputeBackTurnPoseFromBoxToTarget box_pos="{box_pose}"
                                                    box_quat="{box_quat}"
                                                    target_box_offset="-0.6,0,0"
                                                    destination="{target_pose_pre_put}"
                                                    target_pose="{back_target_pose}"/>
                <Fallback>
                  <CloseToDestination pos_error_threshold="0.03"
                                      yaw_error_threshold="5"
                                      destination="{back_target_pose}"
                                      pose_delta="{pose_delta}"/>
                  <SingleStepToDestination destination="{pose_delta}"
                                           step_type="0"/>
                </Fallback>
              </Sequence>
            </TimingDecorator>
            <TimingDecorator name="MoveToPutCMD">
              <Sequence>
                <RetryUntilSuccessful num_attempts="3">
                  <BoxTagOK tag_id="2"
                            box_pos="{box_pose2}"
                            box_quat="{box_quat2}"/>
                </RetryUntilSuccessful>
                <ComputeTargetPose box_pos="{box_pose2}"
                                   box_quat="{box_quat2}"
                                   target_box_offset="{pre_grab_box_offset2}"
                                   target_pose="{target_pose_pre_put}"/>
                <MakePlan target_pose="{target_pose_pre_put}"
                          planed_trajectory="{planed_trajectory2}"/>
                <MoveToDestination planed_trajectory="{planed_trajectory2}"
                                   move_type="1"/>
              </Sequence>
            </TimingDecorator>
            <Sequence>
              <TimingDecorator name="Put">
                <Sequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <BoxTagOK tag_id="2"
                              box_pos="{box_pose2}"
                              box_quat="{box_quat2}"/>
                  </RetryUntilSuccessful>
                  <RetryUntilSuccessful num_attempts="5">
                    <GraspBox box_pos="{box_pose2}"
                              box_quat="{box_quat2}"
                              box_size="{box_size}"
                              box_offset="-0.13,0.0,-0.55"
                              grasp_type="1"/>
                  </RetryUntilSuccessful>
                </Sequence>
              </TimingDecorator>
              <TimingDecorator name="PutBackCMD">
                <Sequence>
                  <ComputeTargetPose box_pos="{box_pose2}"
                                     box_quat="{box_quat2}"
                                     target_box_offset="-1.0,0,0"
                                     target_pose="{target_pose_pre_put}"/>
                  <MakePlan target_pose="{target_pose_pre_put}"
                            planed_trajectory="{planed_trajectory2}"/>
                  <MoveToDestination planed_trajectory="{planed_trajectory2}"
                                     move_type="0"/>
                </Sequence>
              </TimingDecorator>
              <TimingDecorator name="ArmHome">
                <ArmMoveToHomePose/>
              </TimingDecorator>
            </Sequence>
          </SequenceStar>
        </SequenceStar>
      </Sequence>
    </TimingDecorator>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Action ID="ArmMoveToHomePose"
            editable="true"/>
    <Action ID="ArmMoveToReadyPose"
            editable="true"/>
    <Condition ID="BoxTagOK"
               editable="true">
      <input_port name="tag_id"/>
      <output_port name="box_pos"/>
      <output_port name="box_quat"/>
    </Condition>
    <Condition ID="CheckStatusOK"
               editable="true"/>
    <Condition ID="CloseToDestination"
               editable="true">
      <input_port name="pos_error_threshold"/>
      <input_port name="yaw_error_threshold"/>
      <input_port name="destination"/>
      <output_port name="pose_delta"/>
    </Condition>
    <Action ID="ComputeBackTurnPoseFromBoxToTarget"
            editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat"/>
      <input_port name="target_box_offset"/>
      <input_port name="destination"/>
      <output_port name="target_pose"/>
    </Action>
    <Condition ID="ComputeTargetPose"
               editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat"/>
      <input_port name="target_box_offset"/>
      <output_port name="target_pose"/>
    </Condition>
    <Action ID="GraspBox"
            editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat">xyzw</input_port>
      <input_port name="box_size"/>
      <input_port name="box_offset"/>
      <input_port name="grasp_type">0:grasp, 1:put</input_port>
    </Action>
    <Action ID="MakePlan"
            editable="true">
      <input_port name="target_pose"/>
      <output_port name="planed_trajectory"/>
    </Action>
    <Action ID="MoveToDestination"
            editable="true">
      <input_port name="planed_trajectory"/>
      <input_port name="move_type"
                  default="0">int</input_port>
    </Action>
    <Action ID="SingleStepToDestination"
            editable="true">
      <input_port name="destination">x,y,z,yaw</input_port>
      <input_port name="step_type"
                  default="0"/>
    </Action>
    <Decorator ID="TimingDecorator"
               editable="true"/>
  </TreeNodesModel>

</root>
